packages/core/src/anchor/processing-loop.ts

```diff
@@ -70,90 +78,77 @@ export class ProcessingLoop<T> {
       let isDone = false
+      const taskIds = new Set<string>()
       do {
-        try {
-          const curTaskId = taskId
-          await this.#semaphore.acquire().then(async (semRelease) => {
-            this.#logger.verbose(`ProcessingLoop: Fetching next event from source`)
-            const next = await Promise.race([this.source.next(), rejectOnAbortSignal])
-            isDone = next.done
-            if (isDone) return
-            const value = next.value
-            if (value === undefined) {
-              this.#logger.verbose(`No value received in ProcessingLoop, skipping this iteration`)
-              return
-            }
-
-            const task = Promise.race([
-              this.handleValue(value),
-              // eslint-disable-next-line @typescript-eslint/no-empty-function
-              rejectOnAbortSignal.then(() => {}), // swallow return value
-            ])
-              .catch((e) => {
-                this.#logger.err(`Error in ProcessingLoop: ${e}`)
-                this.#abortController.abort('ERROR')
-              })
-              .finally(() => {
-                this.#runningTasks.delete(curTaskId)
-                // Semaphore is released only when the actual work of processing the item is
-                // complete.
-                semRelease()
-              })
-
-            // It's important that the semaphore guards adding the task to the running task set,
-            // so that we won't get more tasks created than the concurrency limit.
-            this.#runningTasks.set(taskId, task)
-          })
-
-          // Mod by MAX_SAFE_INTEGER just in case taskId gets really large and needs to wrap around
-          taskId = (taskId + 1) % Number.MAX_SAFE_INTEGER
-        } catch (e) {
-          this.#logger.err(`Error in ProcessingLoop: ${e}`)
+        this.#taskQueue.add(async () => {
+          this.#logger.verbose(`ProcessingLoop: Fetching next event from source`)
+          const next = await Promise.race([this.source.next(), doneOnAbortSignal])
+          isDone = next.done
+          if (isDone) return
+          const value = next.value
+          if (value === undefined) {
+            this.#logger.verbose(`No value received in ProcessingLoop, skipping this iteration`)
+            return
+          }
+          const uniqueTaskId = this.#toUniqueString(value)
+          if (taskIds.has(uniqueTaskId)) {
+            // We have it processing already
+            return
+          }
+          taskIds.add(uniqueTaskId)
+          try {
+            await Promise.race([this.handleValue(value), waitForAbortSignal])
+          } catch (e) {
+            this.#logger.err(`Error in ProcessingLoop: ${e}`)
+            this.#abortController.abort('ERROR')
+          } finally {
+            taskIds.delete(uniqueTaskId)
+          }
+        })
+        if (this.#taskQueue.size >= this.#taskQueue.concurrency) {
+          await this.#taskQueue.onIdle() // Wait till we process the batch
```

Comment:
doesn't this wait until all the tasks in the queue fully completes all its running tasks?  I think we just want to wait until there's room in the queue again to take on more tasks right?

Reply:
Okay, put together a PR to keep the perf benefit of keeping the set of running tasks as full as possible up to the concurrency limit, while also keeping the cleanups and bug fixes from this PR.  Thankfully the code is much cleaner now as a result of @ukstv's work here, and the TaskQueue makes this a very easy improvement to make.  PTAL: https://github.com/ceramicnetwork/js-ceramic/pull/3167